Update using signal API -> input, computed, output, viewChild

pankajparkar opened the pull request:
Use of below Signal API
input
computed
output
viewChild

pankajparkar commented:
Yet to fix lot of unit tests those were failing

muuvmuuv commented on src/demo/app/examples/data.service.ts:
```diff
@@ -19,9 +19,9 @@ export interface Person {
+	private http = inject(HttpClient);
 
-	constructor(private http: HttpClient) {}
+	private _gitHubAccountsCache = new Map<string, []>();
```
If you really don't want to let anybody use it I would recommend you to use `readonly` and `#` private property. Also for other injects.

pankajparkar replied:
@muuvmuuv  Good idea, but I think this can be separate PR, not to deviate the purpose of PR.. 
Don't want clutter up things in single PR.

muuvmuuv commented on src/ng-option-highlight/lib/ng-option-highlight.directive.ts:
```diff
@@ -1,57 +1,47 @@
 
 	private get _canHighlight() {
-		return this._isDefined(this.term) && this._isDefined(this.label);
+		return this._isDefined(this.term()) && this._isDefined(this.label);
```
Better use `linkedSignal` here or `computed`

pankajparkar replied:
Updated to use computed - https://github.com/ng-select/ng-select/pull/2584/commits/b6bdb8f10f571578de699adb937c94b54ce5761c#diff-d9fd1e2e487941eb7caa372db5de046840dfe57b409b435c906f946a2f583279R28

pankajparkar commented on src/ng-select/lib/ng-option.component.ts:
```diff
@@ -1,57 +1,52 @@
-	ngOnDestroy() {
-		this.stateChange$.complete();
+	private get _label() {
+		return this.elementRef.nativeElement.innerHTML || ''.trim();
```
> Is it also any longer needed to build a state change object, I think this can be simplified with just three computed?

I kind of tried to use computed here like below (not three but only one)


```
host: {
  [innerHTML]: 'innerHTML()'
}
```

**TS**
```
_ = afterNextRender(() => {
    this.label.set(this.element.innerHTML);
    // 1st label highlight triggers from here
    this._highlightLabel();
});

innerHTML = computed(() => {
   if (this.label() !== '')
     return ''; // <-- this leads problem with overriding inner HTML
   // .... rest of the code to modify innerHTML
}) 
```

The challenge over here is, the inner HTML is only available `afterNextRender` (equivalent of `ngAfterViewInit`/`ngAfterContentInit`) hook is triggered. 

So when `computed` 1st time triggered `innerHTML` is empty.. and that leads to problem to handle this situation.. if I return anything, it overrides the HTML, and `afterNextRender` going to read the computed return `value.

For now I feel that imperative way of modifying `innerHTML` suits for this case

pankajparkar commented:
All code review comments are addressed.

@muuvmuuv @pavankjadda please feel free to suggest if I missed out on anything. Thanks much. Cheers

pankajparkar commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -734,60 +735,65 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 		const firstItem = items[0];
-		this.bindLabel = this.bindLabel || this._defaultLabel;
-		this._primitive = isDefined(firstItem) ? !isObject(firstItem) : this._primitive || this.bindLabel === this._defaultLabel;
+		this.bindLabel.set(this.bindLabel() || this._defaultLabel);
```
Changing `bindLabel` to `computed` won't workout as `bindLabel` is already a `model`

pankajparkar commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -500,28 +498,27 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 	setDisabledState(state: boolean): void {
-		this._disabled = state;
-		this._cd.markForCheck();
+		this._disabled.set(state);
```
This has to be applied
Earlier, I was thinking local change detection will trigger after setting signal.
But it only triggered when signal is used on UI, in this case it is not, so manual CD is needed.

pankajparkar commented on src/ng-select/lib/ng-option.component.ts:
```diff
@@ -1,57 +1,52 @@
+
+	constructor() {
+		afterNextRender(() => {
+			if (this._label !== this.label()) {
+				this.label.set(this._label);
```
Linking to previous thread - https://github.com/ng-select/ng-select/pull/2584#discussion_r2210492418

pankajparkar commented on src/ng-select/lib/ng-dropdown-panel.component.ts:
```diff
@@ -37,67 +16,66 @@ const SCROLL_SCHEDULER = typeof requestAnimationFrame !== 'undefined' ? animatio
+	private readonly _scrollablePanel = computed(() => this.scrollElementRef()?.nativeElement);
+	private readonly _contentPanel = computed(() => this.contentElementRef()?.nativeElement);
+
+	private _select: HTMLElement | undefined;
```
Anti pattern in typescript to use `!` hence cannot do this.

pankajparkar commented:
@pavankjadda @muuvmuuv AI code review comments are resolved, please help to re-review again.

Thanks, let me know if you see any problems

muuvmuuv commented on src/ng-option-highlight/lib/ng-option-highlight.directive.ts:
```diff
@@ -1,60 +1,60 @@
-import { AfterViewInit, Directive, ElementRef, Input, OnChanges, Renderer2 } from '@angular/core';
+import { afterNextRender, computed, effect, inject, input, signal, Directive, ElementRef, Renderer2 } from '@angular/core';
+
+function isDefined(value: any) {
```
I am a fan of having Angular logic and anything else separated so utils in some else file but also something for another pr. Also do place REGEX on the root scope because browsers cache/compile the otherwise everytime you call the function

muuvmuuv commented on src/ng-select/lib/ng-select.module.ts:
```diff
@@ -57,6 +57,15 @@ import { DefaultSelectionModelFactory } from './selection-model';
 		NgClearButtonTemplateDirective
 	],
-	providers: [{ provide: SELECTION_MODEL_FACTORY, useValue: DefaultSelectionModelFactory }],
+	providers: provideNgSelect(),
```
Not a fan, every call to a function needs to be avoided if not having a practical function. Pun intended.

pavankjadda commented on src/ng-select/lib/ng-dropdown-panel.component.ts:
```diff
@@ -118,19 +96,17 @@ export class NgDropdownPanelComponent implements OnInit, OnChanges, OnDestroy {
 
 	private get _startOffset() {
-		if (this.markedItem) {
+		const markedItem = this.markedItem();
```
@pankajparkar I see new variables like this in many places. We can just use existing format. Any reason for change?

pankajparkar replied:
Updated - https://github.com/ng-select/ng-select/pull/2584/commits/53cf59e5182d0cbc93ac9608a82ee5b0dc0c44be

pavankjadda commented on src/ng-select/lib/ng-dropdown-panel.component.ts:
```diff
@@ -216,7 +192,11 @@ export class NgDropdownPanelComponent implements OnInit, OnChanges, OnDestroy {
 		this._zone.runOutsideAngular(() => {
-			fromEvent(this.scrollElementRef.nativeElement, 'scroll')
+			const scrollablePanel = this._scrollablePanel();
+			if (!scrollablePanel) {
```
Same here

pavankjadda commented on src/ng-select/lib/ng-dropdown-panel.component.ts:
```diff
@@ -393,15 +373,16 @@ export class NgDropdownPanelComponent implements OnInit, OnChanges, OnDestroy {
 
 	private _appendDropdown() {
-		if (!this.appendTo) {
+		const appendTo = this.appendTo();
```
Same here

pankajparkar replied:
Updated - https://github.com/ng-select/ng-select/pull/2584/commits/53cf59e5182d0cbc93ac9608a82ee5b0dc0c44be

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -278,8 +272,9 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 	get currentPanelPosition(): DropdownPosition {
-		if (this.dropdownPanel) {
-			return this.dropdownPanel.currentPosition;
+		const dropdownPanel = this.dropdownPanel();
```
Nit: I don't we need separate variable here

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -291,27 +286,28 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 
 	private get _isTypeahead() {
-		return this.typeahead && this.typeahead.observed;
+		const typeahead = this.typeahead();
```
Nit: I don't we need separate variable here

pankajparkar replied:
Updated - https://github.com/ng-select/ng-select/pull/2584/commits/53cf59e5182d0cbc93ac9608a82ee5b0dc0c44be

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -328,14 +324,15 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 			this.itemsList.clearSelected();
 		}
 		if (changes.items) {
+			this._itemsAreUsed = true;
```
Why is this added here?

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -376,7 +373,8 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 		const target = $event.target;
 
-		if (this.clearButton && this.clearButton.nativeElement === target) {
+		const clearButton = this.clearButton();
```
Nit: I don't we need separate variable here

pankajparkar replied:
Updated - https://github.com/ng-select/ng-select/pull/2584/commits/53cf59e5182d0cbc93ac9608a82ee5b0dc0c44be

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -604,50 +602,54 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 		let tag;
-		if (isFunction(this.addTag)) {
-			tag = (<AddTagFn>this.addTag)(this.searchTerm);
+		const addTag = this.addTag();
```
Nit: I don't we need separate variable here

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -604,50 +602,54 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 		this.blur();
-		if (this.clearButton) {
-			this.clearButton.nativeElement.focus();
+		const clearButton = this.clearButton();
```
Nit: I don't we need separate variable here

pankajparkar replied:
Updated - https://github.com/ng-select/ng-select/pull/2584/commits/53cf59e5182d0cbc93ac9608a82ee5b0dc0c44be

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -604,50 +602,54 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 	trackByOption = (_: number, item: NgOption) => {
-		if (this.trackByFn) {
-			return this.trackByFn(item.value);
+		const trackByFn = this.trackByFn();
```
Nit: I don't we need separate variable here

pankajparkar replied:
Updated - https://github.com/ng-select/ng-select/pull/2584/commits/53cf59e5182d0cbc93ac9608a82ee5b0dc0c44be

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -604,50 +602,54 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 
 	showNoItemsFound() {
 		const empty = this.itemsList.filteredItems.length === 0;
+		const loading = this.loading();
```
Nit: I don't we need separate variable here

pankajparkar replied:
Updated - https://github.com/ng-select/ng-select/pull/2584/commits/53cf59e5182d0cbc93ac9608a82ee5b0dc0c44be

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -734,60 +736,65 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 			this.itemsList.mapSelectedItems();
 		}
-		if (this.isOpen && isDefined(this.searchTerm) && !this._isTypeahead) {
+		const isOpen = this.isOpen();
```
Nit: I don't we need separate variable here

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -734,60 +736,65 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
-			mapNgOptions(options);
-			handleOptionChange();
-		});
+		this.ngOptionsObservable.pipe(
```
Can you briefly explain why this logic changed?

pankajparkar replied:
Earlier there were 2 subscribes
1. `this.ngOptions.changes.pipe(startWith(this.ngOptions), takeUntil(this._destroy$))`
2. from `handleOptionChange` function.

Instead of 2 subscribe things have been converted to one..

Explaination
```
		this.ngOptionsObservable.pipe(
			startWith(this.ngOptions()),
			takeUntil(this._destroy$),
			tap((options) => {
                                 // 1st subscribe code moved into tap operator
				this.bindLabel.set(this._defaultLabel);
				mapNgOptions(options);
				this._cd.detectChanges();
			}),
                         // second subscribe moved to `handleOptionChange` with switchMap.
                         // why switchMap? to focus on current observable change, 
                         // automatically cancels prev inner obs subscription
			switchMap((options) => handleOptionChange(options))
		).subscribe();
```

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -734,60 +736,65 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 
 	private _isValidWriteValue(value: any): boolean {
-		if (!isDefined(value) || (this.multiple && value === '') || (Array.isArray(value) && value.length === 0)) {
+		const multiple = this.multiple();
```
Nit: I don't we need separate variable here

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -889,13 +897,14 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 		const model = [];
 		for (const item of this.selectedItems) {
-			if (this.bindValue) {
+			const bindValue = this.bindValue();
```
Nit: I don't we need separate variable here

pankajparkar replied:
Updated - https://github.com/ng-select/ng-select/pull/2584/commits/53cf59e5182d0cbc93ac9608a82ee5b0dc0c44be

pavankjadda commented on src/ng-select/lib/ng-select.component.ts:
```diff
@@ -927,43 +936,46 @@ export class NgSelectComponent implements OnDestroy, OnChanges, OnInit, AfterVie
 
 	private _scrollToTag() {
-		if (!this.isOpen || !this.dropdownPanel) {
+		const dropdownPanel = this.dropdownPanel();
```
Nit: I don't we need separate variable here

pavankjadda reviewed (commented):
All right. Great work @pankajparkar 

- Most of the comments of Nit picks. I think we don't need to create separate variables from signals.
- And added few comments about the logic changes
- I did not check demo app changes, I am sure you have taken care of them
If you can address those, we should be able to merge.

pankajparkar commented:
> All right. Great work @pankajparkar
> 
> * Most of the comments of Nit picks. I think we don't need to create separate variables from signals.
> * And added few comments about the logic changes
> * I did not check demo app changes, I am sure you have taken care of them
>   If you can address those, we should be able to merge.

Thanks.
- The reason behind creating separate variable was [type narrowing issue with `signals`](https://github.com/angular/angular/issues/49161), that generally happens when your variable `Type|undefined`, in this case our variable holds mostly a single type.. so that issue doesn't apply here anymore.. Converted all variables to use `this.signal()` way as suggested
- Responded separately
- I would recommend you to double check, just a second eyes.

Thanks for the re-review patiently @muuvmuuv & @pavankjadda

pavankjadda commented:
Looks good. Will review tests and demo as sanity check.

muuvmuuv commented on src/ng-select/lib/ng-select.component.ts:
This new GH view for re-viewing is hard... I wanted to comment on line 355, you can replace all ngOnDestroy logic that includes observables with takeUntilDestroyed from rxjs-interap Angular packages and use the component DestroyRef.

muuvmuuv replied:
Yeah ofc :) just a thought that came up while reviewing. lgtm

pankajparkar commented:
@pavankjadda I had raise a PR for the bugs.. and addressed both them..

Can you please let me know what are `has lot of bugs too`? I currently testing it thoroughly locally.. but curious if you have any specific that I can them properly. Thanks

Hypercubed commented:
FYI... these changes in 15.2 are breaking for some (at least one) downstream library.  See https://github.com/ng-matero/extensions/issues/438#issuecomment-3152812973 .  I see these were reverted on in v20 major, but doesn't help for anyone depending on `^15` expecting no breaking changes.   I suggest a 15.2.1 to revert this.
